system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -63,70 +101,422 @@ class CLibCompiledXpath : public CInterface, public ICompiledXpath
+        exsltSetsXpathCtxtRegister(m_xpathContext, (xmlChar*)"set");
+        exsltStrXpathCtxtRegister(m_xpathContext, (xmlChar*)"str");
     }
+
     ~CLibXpathContext()
     {
+        for (XPathInputMap::iterator it=provided.begin(); it!=provided.end(); ++it)
+            it->second->Release();
         xmlXPathFreeContext(m_xpathContext);
         xmlFreeDoc(m_xmlDoc);
     }
 
+    void pushLocation()
+    {
+        WriteLockBlock wblock(m_rwlock);
+        saved.emplace_back(XpathContextState(m_xpathContext));
+    }
+
+    void setLocation(xmlDocPtr doc, xmlNodePtr node, int contextSize, int proximityPosition)
+    {
+        WriteLockBlock wblock(m_rwlock);
+        m_xpathContext->doc = doc;
+        m_xpathContext->node = node;
+        m_xpathContext->contextSize = contextSize;
+        m_xpathContext->proximityPosition = proximityPosition;
+    }
+
+    void setLocation(xmlXPathContextPtr ctx)
+    {
+        WriteLockBlock wblock(m_rwlock);
+        m_xpathContext->doc = ctx->doc;
+        m_xpathContext->node = ctx->node;
+        m_xpathContext->contextSize = ctx->contextSize;
+        m_xpathContext->proximityPosition = ctx->proximityPosition;
+    }
+
+    void popLocation()
+    {
+        saved.back().restore(m_xpathContext);
+        saved.pop_back();
+    }
+
+    void beginScope(const char *name) override
+    {
+        WriteLockBlock wblock(m_rwlock);
+        scopes.emplace_back(new CLibXpathScope(name));
+    }
+
+    void endScope() override
+    {
+        WriteLockBlock wblock(m_rwlock);
+        if (scopes.size()>1) //preserve root scope
+            scopes.pop_back();
+    }
+
     static void tableScanCallback(void *payload, void *data, xmlChar *name)
     {
         DBGLOG("k/v == [%s,%s]\n", (char *) name, (char *) payload);
     }
 
+    virtual void registerNamespace(const char *prefix, const char * uri) override
+    {
+        if (m_xpathContext)
+        {
+            WriteLockBlock wblock(m_rwlock);
+            xmlXPathRegisterNs(m_xpathContext, (const xmlChar *) prefix, (const xmlChar *) uri);
+        }
+    }
+
+    virtual void registerFunction(const char *xmlns, const char * name, void *f) override
+    {
+        if (m_xpathContext)
+        {
+            WriteLockBlock wblock(m_rwlock);
+            xmlXPathRegisterFuncNS(m_xpathContext, (const xmlChar *) name, (const xmlChar *) xmlns, (xmlXPathFunction) f);
+        }
+    }
+
+    virtual void setUserData(void *userdata) override
+    {
+        if (m_xpathContext)
+        {
+            WriteLockBlock wblock(m_rwlock);
+            m_xpathContext->userData = userdata;
+        }
+    }
+
+    virtual void *getUserData() override
+    {
+        if (m_xpathContext)
+        {
+            ReadLockBlock wblock(m_rwlock);
+            return m_xpathContext->userData;
+        }
+        return nullptr;
+    }
+
     virtual const char * getXpath() override
     {
+        ReadLockBlock wblock(m_rwlock);
```

Comment:
I was going to discuss requiring any additional threads to have their own contexts and removing sync.  but that's a different PR.  I'll check and if this value isn't used we can just remove it.

Reply:
Additional threads should have their own contexts, because they should have their own variables and will likely be working at different locations in the document. I think synchronization remains an issue if multiple contexts share a document, especially when at least one of the contexts is used to write to the document.

Reply:
So far we never write to the xpath document within a script.  If we started to use libxml2 on the right hand side that context would be writable.  but I doubt we'd want more than one thread operating on the same target document at the same time.  I know we discussed theoretical multi-threading but we need to be careful and probably error on the side of simplicity.  In any case that's future work